analyzer/java/com/engflow/bazel/invocation/analyzer/bazelprofile/BazelProfile.java

```diff
@@ -77,69 +77,77 @@ public static BazelProfile createFromPath(String path) throws IllegalArgumentExc
 
   public static BazelProfile createFromInputStream(InputStream inputStream)
       throws IllegalArgumentException {
-    JsonObject bazelProfile;
-    try {
-      bazelProfile = JsonParser.parseReader(new InputStreamReader(inputStream)).getAsJsonObject();
-    } catch (IllegalStateException e) {
-      throw new IllegalArgumentException("Could not parse Bazel profile.", e);
-    }
-    return new BazelProfile(bazelProfile);
+    return new BazelProfile(new JsonReader(new InputStreamReader(inputStream)));
   }
 
   private final BazelVersion bazelVersion;
   private final Map<String, String> otherData = new HashMap<>();
   private final Map<ThreadId, ProfileThread> threads = new HashMap<>();
 
-  private BazelProfile(JsonObject profile) {
-    if (!profile.has(TraceEventFormatConstants.SECTION_OTHER_DATA)
-        || !profile.has(TraceEventFormatConstants.SECTION_TRACE_EVENTS)) {
-      throw new IllegalArgumentException(
-          String.format(
-              "Invalid profile, JSON file missing \"%s\" and/or \"%s\"",
-              TraceEventFormatConstants.SECTION_OTHER_DATA,
-              TraceEventFormatConstants.SECTION_TRACE_EVENTS));
-    }
+  private BazelProfile(JsonReader profileReader) {
```

Comment:
Stylistically, it is not recommended to do significant work in a constructor. It would be better to refactor this code and do the parsing in a static method or a builder class, and then construct the object with just the required fields. While at it, it may also make sense to use immutable maps for the fields.

Reply:
(Not this PR)

Reply:
Ack.